src/Builder/FrontendDialectTransformer.cpp

```diff
@@ -983,16 +983,12 @@ class FrontendGenImpl {
 
       // Update types:
       for (int i = 0; i < n.output_size(); ++i) {
-        std::unique_ptr<onnx::TypeProto> p =
-            std::make_unique<onnx::TypeProto>(*node_ctx.getOutputType(i));
-        typeMap[n.output(i)] = p.get();
-        types.push_back(std::move(p));
+        const std::string &output_name = n.output(i);
+        onnx_type_map.AddMapping(output_name, *node_ctx.getOutputType(i));
+        typeMap[output_name] = const_cast<onnx::TypeProto *>(
+            onnx_type_map.GetByOnnxName(output_name));
       }
     }
-
```

Comment:
I noticed that the modification moved the AddMapping from the outside of node loop into the node loop. Could one element of typeMap be assigned with different value inside the node loop? I guess not since an output tensor should be defined only once by onnx semantic, and our shape inference updates the type of the output of the current node. I somehow understand why Tian wrote the code in this way.

Reply:
in theory the old code would allow an attempted insertion of an output into typeMap to be ignored if the name was already mapped in typeMap, but I don't think we need to support that because if we look at how the same names are entered into frontend_symbols_, they are entered much the same way with calls to AddMapping (via BindOnnxName and ImportNode+buildOutputAndOperation)

in other words, in practice the names added are distinct (because otherwise frontend_symbols_.AddMapping would break somewhere else) so it works to call onnx_type_map.AddMapping directly without going through typeMap